File: app/src/main/java/com/sg/moviesindex/model/Movie.java
package com.sg.moviesindex.model;

import java.util.ArrayList;
import java.util.List;

import android.annotation.SuppressLint;
import android.os.Parcel;
import android.os.Parcelable;
import com.sg.moviesindex.BR;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;
import androidx.annotation.NonNull;
import androidx.databinding.BaseObservable;
import androidx.databinding.Bindable;
import androidx.recyclerview.widget.DiffUtil;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.PrimaryKey;

@Entity(tableName = "favourite_movies")
public class Movie extends BaseObservable implements Parcelable
{

    @ColumnInfo(name = "vote_count")
    @SerializedName("vote_count")
    @Expose
    private Integer voteCount;

    @ColumnInfo(name = "id")
    @PrimaryKey(autoGenerate = true)
    @SerializedName("id")
    @Expose
    private Integer id;

    @ColumnInfo(name = "video")
    @SerializedName("video")
    @Expose
    private Boolean video;

    @ColumnInfo(name = "vote_average")
    @SerializedName("vote_average")
    @Expose
    private Double voteAverage;

    @ColumnInfo(name = "title")
    @SerializedName("title")
    @Expose
    private String title;

    @ColumnInfo(name = "popularity")
    @SerializedName("popularity")
    @Expose
    private Double popularity;

    @ColumnInfo(name = "poster_path")
    @SerializedName("poster_path")
    @Expose
    private String posterPath;


    @ColumnInfo(name = "original_language")
    @SerializedName("original_language")
    @Expose
    private String originalLanguage;

    @ColumnInfo(name = "original_title")
    @SerializedName("original_title")
    @Expose
    private String originalTitle;



    @Ignore
    @SerializedName("genre_ids")
    @Expose
    private List<Integer> genreIds = new ArrayList<>();



    @ColumnInfo(name = "backdrop_path")
    @SerializedName("backdrop_path")
    @Expose
    private String backdropPath;

    @ColumnInfo(name = "adult")
    @SerializedName("adult")
    @Expose
    private Boolean adult;

    @ColumnInfo(name = "overview")
    @SerializedName("overview")
    @Expose
    private String overview;

    @ColumnInfo(name = "release_date")
    @SerializedName("release_date")
    @Expose
    private String releaseDate;


    @Ignore
    public final static Parcelable.Creator<Movie> CREATOR = new Creator<Movie>() {

        @Ignore
        @SuppressWarnings({
                "unchecked"
        })
        public Movie createFromParcel(Parcel in) {
            return new Movie(in);
        }

        @Ignore
        public Movie[] newArray(int size) {
            return (new Movie[size]);
        }

    };

    @Ignore
    protected Movie(Parcel in) {
        this.voteCount = ((Integer) in.readValue((Integer.class.getClassLoader())));
        this.id = ((Integer) in.readValue((Integer.class.getClassLoader())));
        this.video = ((Boolean) in.readValue((Boolean.class.getClassLoader())));
        this.voteAverage = ((Double) in.readValue((Double.class.getClassLoader())));
        this.title = ((String) in.readValue((String.class.getClassLoader())));
        this.popularity = ((Double) in.readValue((Double.class.getClassLoader())));
        this.posterPath = ((String) in.readValue((String.class.getClassLoader())));
        this.originalLanguage = ((String) in.readValue((String.class.getClassLoader())));
        this.originalTitle = ((String) in.readValue((String.class.getClassLoader())));
        in.readList(this.genreIds, (java.lang.Integer.class.getClassLoader()));
        this.backdropPath = ((String) in.readValue((String.class.getClassLoader())));
        this.adult = ((Boolean) in.readValue((Boolean.class.getClassLoader())));
        this.overview = ((String) in.readValue((String.class.getClassLoader())));
        this.releaseDate = ((String) in.readValue((String.class.getClassLoader())));
    }

    @Ignore
    public Movie() {
    }

    public Movie(Integer voteCount, Integer id, Boolean video, Double voteAverage, String title, Double popularity, String posterPath, String originalLanguage, String originalTitle, String backdropPath, Boolean adult, String overview, String releaseDate) {
        this.voteCount = voteCount;
        this.id = id;
        this.video = video;
        this.voteAverage = voteAverage;
        this.title = title;
        this.popularity = popularity;
        this.posterPath = posterPath;
        this.originalLanguage = originalLanguage;
        this.originalTitle = originalTitle;
        this.backdropPath = backdropPath;
        this.adult = adult;
        this.overview = overview;
        this.releaseDate = releaseDate;

    }

    @Bindable
    public Integer getVoteCount() {
        return voteCount;
    }

    public void setVoteCount(Integer voteCount) {
        this.voteCount = voteCount;
        notifyPropertyChanged(BR.voteCount);
    }

    @Bindable
    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
        notifyPropertyChanged(BR.id);
    }

    @Bindable
    public Boolean getVideo() {
        return video;
    }

    public void setVideo(Boolean video) {
        this.video = video;
        notifyPropertyChanged(BR.video);
    }

    @Bindable
    public Double getVoteAverage() {
        return voteAverage;
    }

    public void setVoteAverage(Double voteAverage) {
        this.voteAverage = voteAverage;

        notifyPropertyChanged(BR.voteAverage);
    }

    @Bindable
    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
        notifyPropertyChanged(BR.title);
    }

    @Bindable
    public Double getPopularity() {
        return popularity;
    }

    public void setPopularity(Double popularity) {
        this.popularity = popularity;
        notifyPropertyChanged(BR.popularity);
    }

    @Bindable
    public String getPosterPath() {
        return posterPath;
    }

    public void setPosterPath(String posterPath) {
        this.posterPath = posterPath;
        notifyPropertyChanged(BR.posterPath);
    }

    @Bindable
    public String getOriginalLanguage() {
        return originalLanguage;
    }

    public void setOriginalLanguage(String originalLanguage) {
        this.originalLanguage = originalLanguage;
        notifyPropertyChanged(BR.originalLanguage);
    }

    @Bindable
    public String getOriginalTitle() {
        return originalTitle;
    }

    public void setOriginalTitle(String originalTitle) {
        this.originalTitle = originalTitle;
        notifyPropertyChanged(BR.originalTitle);
    }

    @Ignore
    public List<Integer> getGenreIds() {
        return genreIds;
    }

    @Ignore
    public void setGenreIds(List<Integer> genreIds) {
        this.genreIds = genreIds;
    }

    @Bindable
    public String getBackdropPath() {
        return backdropPath;
    }

    public void setBackdropPath(String backdropPath) {
        this.backdropPath = backdropPath;
        notifyPropertyChanged(BR.backdropPath);
    }

    @Bindable
    public Boolean getAdult() {
        return adult;
    }

    public void setAdult(Boolean adult) {
        this.adult = adult;
        notifyPropertyChanged(BR.adult);
    }

    @Bindable
    public String getOverview() {
        return overview;
    }

    public void setOverview(String overview) {
        this.overview = overview;
        notifyPropertyChanged(BR.overview);
    }

    @Bindable
    public String getReleaseDate() {
        return releaseDate;
    }

    public void setReleaseDate(String releaseDate) {
        this.releaseDate = releaseDate;
        notifyPropertyChanged(BR.releaseDate);
    }

    @Ignore
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(voteCount);
        dest.writeValue(id);
        dest.writeValue(video);
        dest.writeValue(voteAverage);
        dest.writeValue(title);
        dest.writeValue(popularity);
        dest.writeValue(posterPath);
        dest.writeValue(originalLanguage);
        dest.writeValue(originalTitle);
        dest.writeList(genreIds);
        dest.writeValue(backdropPath);
        dest.writeValue(adult);
        dest.writeValue(overview);
        dest.writeValue(releaseDate);
    }

    @Ignore
    public int describeContents() {
        return 0;
    }

    public static final DiffUtil.ItemCallback<Movie> callback=new DiffUtil.ItemCallback<Movie>() {
        @Override
        public boolean areItemsTheSame(@NonNull Movie oldItem, @NonNull Movie newItem) {
            return oldItem.id.equals(newItem.id);
        }

        @SuppressLint("DiffUtilEquals")
        @Override
        public boolean areContentsTheSame(@NonNull Movie oldItem, @NonNull Movie newItem) {
            return oldItem.adult == newItem.adult &&
                    oldItem.video == newItem.video &&
                    oldItem.backdropPath.equals(newItem.backdropPath) &&
                    oldItem.genreIds == newItem.genreIds &&
                    oldItem.id.equals(newItem.id) &&
                    oldItem.originalLanguage.equals(newItem.originalLanguage) &&
                    oldItem.originalTitle.equals(newItem.originalTitle) &&
                    oldItem.overview.equals(newItem.overview) &&
                    oldItem.popularity.equals(newItem.popularity) &&
                    oldItem.posterPath.equals(newItem.posterPath) &&
                    oldItem.releaseDate.equals(newItem.releaseDate) &&
                    oldItem.title.equals(newItem.title) &&
                    oldItem.voteAverage.equals(newItem.voteAverage) &&
                    oldItem.voteCount.equals(newItem.voteCount);
        }
    };
}

File: app/src/main/java/com/sg/moviesindex/data/local/Movie.kt
package com.sg.moviesindex.data.local

import android.annotation.SuppressLint
import android.os.Parcelable
import androidx.recyclerview.widget.DiffUtil
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import com.sg.moviesindex.data.remote.Cast
import com.sg.moviesindex.data.remote.Genre
import com.sg.moviesindex.data.remote.Review
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import kotlinx.parcelize.IgnoredOnParcel
import kotlinx.parcelize.Parcelize

@Entity(tableName = "favourite_movies")
@Parcelize
@JsonClass(generateAdapter = true)
data class Movie(
  @ColumnInfo(name = "adult")
  @Json(name = "adult")
  var adult: Boolean? = null,
  @ColumnInfo(name = "backdrop_path")
  @Json(name = "backdrop_path")
  var backdropPath: String? = null,
  @ColumnInfo(name = "budget")
  @Json(name = "budget")
  var budget: Long? = null,
  @ColumnInfo(name = "genres")
  @Json(name = "genres")
  var genres: List<Genre> = emptyList(),
  @PrimaryKey(autoGenerate = true)
  @ColumnInfo(name = "id")
  @Json(name = "id")
  var id: Long? = null,
  @ColumnInfo(name = "imdb_id")
  @Json(name = "imdb_id")
  var imdbId: String? = null,
  @ColumnInfo(name = "original_language")
  @Json(name = "original_language")
  var originalLanguage: String? = null,
  @ColumnInfo(name = "original_title")
  @Json(name = "original_title")
  var originalTitle: String? = null,
  @ColumnInfo(name = "overview")
  @Json(name = "overview")
  var overview: String? = null,
  @ColumnInfo(name = "popularity")
  @Json(name = "popularity")
  var popularity: Double? = null,
  @ColumnInfo(name = "poster_path")
  @Json(name = "poster_path")
  var posterPath: String? = null,
  @ColumnInfo(name = "release_date")
  @Json(name = "release_date")
  var releaseDate: String? = null,
  @ColumnInfo(name = "revenue")
  @Json(name = "revenue")
  var revenue: Long? = null,
  @ColumnInfo(name = "runtime")
  @Json(name = "runtime")
  var runtime: Long? = null,
  @ColumnInfo(name = "status")
  @Json(name = "status")
  var status: String? = null,
  @ColumnInfo(name = "tagline")
  @Json(name = "tagline")
  var tagline: String? = null,
  @ColumnInfo(name = "title")
  @Json(name = "title")
  var title: String? = null,
  @ColumnInfo(name = "video")
  @Json(name = "video")
  var video: Boolean? = null,
  @ColumnInfo(name = "vote_average")
  @Json(name = "vote_average")
  var voteAverage: Double? = null,
  @ColumnInfo(name = "casts_list")
  var castsList: List<Cast> = emptyList(),
  @ColumnInfo(name = "reviews_list")
  var reviewsList: List<Review> = emptyList(),
) : Parcelable {
  @IgnoredOnParcel
  @Ignore
  @Json(name = "genre_ids")
  var genreIds: List<Int> = emptyList()

  override fun toString(): String = "Movie(title='$title', originalLanguage='$originalLanguage')"

  companion object {
    val callback =
      object : DiffUtil.ItemCallback<Movie>() {
        override fun areItemsTheSame(
          oldItem: Movie,
          newItem: Movie,
        ): Boolean = oldItem.id == newItem.id

        @SuppressLint("DiffUtilEquals")
        override fun areContentsTheSame(
          oldItem: Movie,
          newItem: Movie,
        ): Boolean = oldItem == newItem
      }
  }
}
